### PR TITLE
Add backend for user authentication

### PR DESCRIPTION
**Pull request by preciousonah**
**Description**

- Add some components for the backend authentication, store user information in the database, add routes to and from the login/logout and sign up page. I will start working on posting in next PR

**Milestones**

- Add a functionality that allows users to  login, logout and sign up into the account
- Add functionality that only allows existing users to login
- Add session management to store the user’s info
- Add get request to retrieve the user’s info and post request to add to the database
- Use hashing to encrypt user password

**Things i could do better:**

- Forgot to use var to define hex colors, will do that in next PR
- Remove the Vercel deployment error. I’m working on that
- Use gitignore to stop tracking the node_modules , will add that in next PR
- Remove all the console.log statements
**NOTE**: I have added a new commit that addresses the changes above. Also, there is some code that got pushed in with the because there were uncommitted changes in my last push. I will fix this error in future pull requests and ensure I push all the committed changes next time. Here is the recent commit I made:
https://github.com/preciousonah/Fashion-Messaging-Capstone-Project-/pull/11/commits/3daf08e016e125ef963cbb0b5aa000e49b144d8d

Here’s a preview of my website below:
Screenshot 2023-07-14 at 2 30 26 PM
Image 7-14-23 at 2 31 PM
Image 7-14-23 at 2 29 PM

### CODE REVIEW

File: fashion-messaging-app-api/server.js [hunk 1]
```diff
@@ -14,7 +14,8 @@ app.use(cors({
   credentials: true
 }));
 app.use(express.json()); 
-app.use(morgan())
+app.use(morgan('combined'));
```
Comment (preciousonah) [inline fashion-messaging-app-api/server.js]: I changed this because I got an error that Morgan deprecated because I didn't specify a format so I added 'combined' to fix the bug

File: fashion-messaging-app-ui/src/UserContext.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+import { createContext } from 'react';
+
+export const UserContext = createContext();
```
Comment (preciousonah) [inline fashion-messaging-app-ui/src/UserContext.js:3]: This was part of the code that was not added to the last file I pushed

File: fashion-messaging-app-ui/src/main.jsx [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import React from "react"
 import ReactDOM from "react-dom/client"
 import App from "./App.jsx"
-import "./index.css"
+import "./global.css"
```
Comment (preciousonah) [inline fashion-messaging-app-ui/src/main.jsx:4]: I changed this because it has all my color variables for my css so it has more specific naming convention

File: fashion-messaging-app-api/routes/users.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    req.session.user = newUser;
+    res.json({ user: newUser });
+  } catch (error) {
+    console.error(error);
+    res.status(500).json({ error: 'Server error' });
+  }
+});
+
+} catch (error) {
+  console.error(error);}
```
Comment (speltzka) [inline fashion-messaging-app-api/routes/users.js:33]: Maybe try looking into VSCode code formatter options that can automatically fix things like having the curly braces on the correct line. For example the curly brace currently on Line 33 should be on line 34 instead

File: fashion-messaging-app-api/routes/users.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  const { username, password } = req.body;
+
+  try {
+    const user = await User.findOne({ where: { username } });
+
+    if (!user) {
+      return res.status(401).json({ error: 'Invalid username or password' });
+    }
+    const isValidPassword = await bcrypt.compare(password, user.password);
+
```
Comment (speltzka) [inline fashion-messaging-app-api/routes/users.js:45]: Good job handling all these errors!

File: fashion-messaging-app-api/server.js [hunk 6]
```diff
@@ -1,11 +1,86 @@
...
+app.use(
+  session({
+    secret: 'your-secret-key',
+    resave: false,
+    saveUninitialized: false,
+    store: sessionStore,
+    cookie: {
+      sameSite: false,
+      secure: false,
+      expires: new Date(Date.now() + (365 * 24 * 60 * 60 * 1000))
```
Comment (speltzka) [inline fashion-messaging-app-api/server.js:34]: Haha what is this expiration date? Would be good to define a variable that explains what (365 * 24 * 60 * 60 * 1000) represents

File: fashion-messaging-app-api/server.js [hunk 7]
```diff
@@ -1,11 +1,86 @@
...
+
+    const postWithUser = await Post.findOne({
+      where: { id: post.id },
+      include: [{ model: User, as: 'user' }]
+    });
+
+    res.status(201).json(postWithUser);
+  } catch (err) {
+    res.status(500).json({ message: err.message });
+  }
```
Comment (speltzka) [inline fashion-messaging-app-api/server.js:74]: It seems like all this code relates to posts instead of Authentication. Please make sure to not include unrelated code in your PRs. Everything included should relate to the stated purpose of the pull request (in this case "Add backend for user authentication")

File: fashion-messaging-app-api/updatePassword.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    for (const user of users) {
+      const hashedPassword = await bcrypt.hash(user.password, 10);
+      await user.update({ password: hashedPassword });
+    }
+  } catch (error) {
+    console.error('User migration failed:', error);
+  }
+}
+
+migrateUsers();
```
Comment (speltzka) [inline fashion-messaging-app-api/updatePassword.js:16]: It seems like this whole file is for changing your password. I don't think you are using this and I also don't think it relates directly to User Authentication. Try to make sure not to include unrelated or unused code

File: fashion-messaging-app-ui/src/components/SignupForm/SignupForm.css [hunk 9]
```diff
@@ -10,7 +10,7 @@
       flex-direction: column;
       width: 300px;
       padding: 20px;
-      border: 1px solid #ddd;
+      border: 1px solid var(  --white);
```
Comment (speltzka) [inline fashion-messaging-app-ui/src/components/SignupForm/SignupForm.css:13]: Thanks for defining the colors!

File: fashion-messaging-app-ui/src/components/SignupForm/SignupForm.jsx [hunk 10]
```diff
@@ -1,12 +1,44 @@
...
+        credentials: 'include'
+      });
+
+      if (response.ok) {
+        const data = await response.json();
+        const loggedInUser = data.user;
+
+        setUsername('');
+        setEmail('');
+        setPassword('');
```
Comment (speltzka) [inline fashion-messaging-app-ui/src/components/SignupForm/SignupForm.jsx:31]: Why do you need these 3 lines?
